packages/nouns-contracts/contracts/libs/Noracle.sol

```diff
@@ -0,0 +1,140 @@
+        uint32 cardinality;
+        uint32 cardinalityNext;
+    }
+
+    function initialize(NoracleState storage self) internal {
+        if (self.cardinality > 0) revert AlreadyInitialized();
+
+        self.cardinality = 1;
+        self.cardinalityNext = 1;
+        warmUpObservation(self.observations[0]);
+    }
+
+    function write(
+        NoracleState storage self,
+        uint32 blockTimestamp,
+        uint16 nounId,
+        uint48 amount,
+        address winner
+    ) internal {
+        uint32 cardinalityNext = self.cardinalityNext;
+        if (cardinalityNext == 0) revert NotInitialized();
+
+        // if the conditions are right, we can bump the cardinality
+        uint32 currentIndex = self.index;
+        if (cardinalityNext > self.cardinality && currentIndex == (self.cardinality - 1)) {
+            self.cardinality = cardinalityNext;
+        }
+
+        uint32 newIndex = (currentIndex + 1) % self.cardinality;
+        self.index = newIndex;
+        self.observations[newIndex] = Observation({
+            blockTimestamp: blockTimestamp,
+            nounId: nounId,
+            amount: amount,
+            winner: winner
+        });
+    }
+
+    function grow(NoracleState storage self, uint32 next) internal returns (uint32) {
+        uint32 current = self.cardinalityNext;
+        if (current == 0) revert NotInitialized();
+
+        // no-op if the passed next value isn't greater than the current next value
+        if (next <= current) return current;
+
+        // store in each slot to prevent fresh SSTOREs
+        // this data will not be used because the initialized boolean is still false
+        for (uint32 i = current; i < next; i++) {
+            warmUpObservation(self.observations[i]);
+        }
+
+        return self.cardinalityNext = next;
+    }
+
+    function observe(NoracleState storage self, uint32 auctionCount)
+        internal
+        view
+        returns (Observation[] memory observations)
+    {
+        uint32 cardinality = self.cardinality;
+        if (auctionCount > cardinality) revert AuctionCountOutOfBounds(auctionCount, cardinality);
+
+        uint32 index = self.index;
+        observations = new Observation[](auctionCount);
+
+        uint32 observationsCount = 0;
+        while (observationsCount < auctionCount) {
+            uint32 checkIndex = (index + (cardinality - observationsCount)) % cardinality;
+            Observation storage obs = self.observations[checkIndex];
+
+            // when cardinality has been increased, and not used up yet, there are uninitialized items
+            // this loop reads from index backwards, so when it hits an uninitialized item
+            // it means all items from there backwards until index are uninitialized, so we should break.
+            if (!initialized(obs)) break;
+
+            observations[observationsCount] = obs;
+            ++observationsCount;
+        }
+
+        if (auctionCount > observationsCount) {
```

Comment:
wondering if it's better to drop this block and return some extra zero array members.
or if there's a better way to deal with this.
still thinking about it

Reply:
yes, I am also not very comfortable with the current state
AFAIK solidity still doesn't support dynamic-size arrays
and we don't have enough information at the beginning to know exactly how many items to allocate

btw, UniV3's observe function is also not very cheap on gas
https://github.com/Uniswap/v3-core/blob/main/contracts/libraries/Oracle.sol#L314

Reply:
Should be able to "shorten" by modifying the length:

```solidity
if (auctionCount > observationsCount) {
    assembly {
        mstore(observations, observationsCount)
    }
}
```

Reply:
great feedback, fixed!

Reply:
@solimander nice hack :)